In [1]:
import numpy as np
import pandas as pd

<h1><b>1. Khám phá dữ liệu (đan xen với tiền xử lý dữ liệu)</b></h1>

### Đọc dữ liệu từ tập tin (file)

Đầu tiên, đọc dữ liệu từ file "kaggle_survey_2022_responses.csv" và lưu kết quả vào DataFrame df.

In [2]:
df_1 = pd.read_csv("kaggle_survey_2022_responses.csv")
df_1.head(5)

c:\Users\KHUONG\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py:3378: DtypeWarning: Columns (0,208,225,255,257,260,270,271,277) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Duration (in seconds)                           Q2  \
0  Duration (in seconds)  What is your age (# years)?   
1                    121                        30-34   
2                    462                        30-34   
3                    293                        18-21   
4                    851                        55-59   

                                       Q3  \
0  What is your gender? - Selected Choice   
1                                     Man   
2                                     Man   
3                                     Man   
4                                     Man   

                                          Q4  \
0  In which country do you currently reside?   
1                                      India   
2                                    Algeria   
3                                      Egypt   
4                                     France   

                                                  Q5  \
0  Are you currently a student? (high school, uni...   
1                                                 No   
2                                                 No   
3                                                Yes   
4                                                 No   

                                                Q6_1  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           Coursera   
4                                           Coursera   

                                                Q6_2  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                edX   
4                                                NaN   

                                                Q6_3  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                               Kaggle Learn Courses   

                                                Q6_4  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           DataCamp   
4                                                NaN   

                                                Q6_5  ...  \
0  On which platforms have you begun or completed...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

                                               Q44_3  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                               Q44_4  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                    Kaggle (notebooks, forums, etc)   
4                    Kaggle (notebooks, forums, etc)   

                                               Q44_5  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Course Forums (forums.fast.ai, Coursera forums...   

                                     

In [3]:
df_1.shape

(23998, 296)

## Tiền xử lý dữ liệu để đưa về dạng có thể phân tích

DÒng đầu tiên là câu hỏi, ta sẽ lấy dòng này lưu vào `dic_question` và xoá dòng này đi

In [4]:
dic_question = dict(df_1.iloc[0,:])
df_1 = df_1.drop(0)

Chuyển `dic_question` thành `new_dic` bằng việc bỏ những câu hỏi trùng nhau

In [5]:
new_dic = dict()
for i in dic_question.keys():
    i1 = i + ' '
    
    if '_1 ' in i1:
        if "?" in dic_question[i]:
            new_dic[i[0:i.index("_")]] = dic_question[i][0:dic_question[i].index('?')+1]
        elif ":" in dic_question[i]: 
            new_dic[i[0:i.index("_")]] = dic_question[i][0:dic_question[i].index(':')]
        else:
            new_dic[i[0:i.index("_")]] = dic_question[i]
    elif '_' not in i:
        if "?" in dic_question[i]:
            new_dic[i] = dic_question[i][0:dic_question[i].index('?')+1]
        elif ":" in dic_question[i]: 
            new_dic[i] = dic_question[i][0:dic_question[i].index(':')]
        else:
            new_dic[i] = dic_question[i]
new_dic['Duration (in seconds)'] = 'Duration'
new_dic

{'Duration (in seconds)': 'Duration',
 'Q2': 'What is your age (# years)?',
 'Q3': 'What is your gender?',
 'Q4': 'In which country do you currently reside?',
 'Q5': 'Are you currently a student?',
 'Q6': 'On which platforms have you begun or completed data science courses?',
 'Q7': 'What products or platforms did you find to be most helpful when you first started studying data science?',
 'Q8': 'What is the highest level of formal education that you have attained or plan to attain within the next 2 years?',
 'Q9': 'Have you ever published any academic research (papers, preprints, conference proceedings, etc)?',
 'Q10': 'Did your research make use of machine learning?',
 'Q11': 'For how many years have you been writing code and/or programming?',
 'Q12': 'What programming languages do you use on a regular basis?',
 'Q13': "Which of the following integrated development environments (IDE's) do you use on a regular basis?",
 'Q14': 'Do you use any of the following hosted notebook products?

Vì mục đích của nhóm là phân tích ở 4 quốc giá là **Japan, China, South Korea, Viet Nam** nên sẽ chỉ lấy dưng dòng thuộc 4 quốc gia này.

In [6]:
df_1 = df_1[(df_1['Q4'] == "Japan") | (df_1['Q4'] == "China") | (df_1['Q4'] == "South Korea") | (df_1['Q4'] == "Viet Nam")]
df_1.head(5)

Duration (in seconds)     Q2     Q3           Q4   Q5 Q6_1 Q6_2  \
11                   470  55-59    Man  South Korea   No  NaN  NaN   
22                   174  25-29  Woman        China  Yes  NaN  NaN   
36                   506  25-29    Man        Japan   No  NaN  NaN   
71                   283  22-24  Woman        China  Yes  NaN  NaN   
72                   492  40-44    Man        Japan   No  NaN  NaN   

                    Q6_3 Q6_4 Q6_5  ... Q44_3  \
11                   NaN  NaN  NaN  ...   NaN   
22  Kaggle Learn Courses  NaN  NaN  ...   NaN   
36                   NaN  NaN  NaN  ...   NaN   
71                   NaN  NaN  NaN  ...   NaN   
72  Kaggle Learn Courses  NaN  NaN  ...   NaN   

                              Q44_4 Q44_5 Q44_6 Q44_7 Q44_8 Q44_9 Q44_10  \
11                              NaN   NaN   NaN   NaN   NaN   NaN    NaN   
22                              NaN   NaN   NaN   NaN   NaN   NaN    NaN   
36  Kaggle (notebooks, forums, etc)   NaN   NaN   NaN   NaN   NaN    NaN   
71                              NaN   NaN   NaN   NaN   NaN   NaN    NaN   
72  Kaggle (notebooks, forums, etc)   NaN   NaN   NaN   NaN   NaN    NaN   

   Q44_11 Q44_12  
11    NaN    NaN  
22    NaN    NaN  
36    NaN    NaN  
71    NaN  Other  
72    NaN    NaN  

[5 rows x 296 columns]

Columnss có 2 dạng, loại 1 là câu hỏi trả lời đáp án, loại 2 là những câu hỏi _1, _2, ... là dạng catogerial nên sẽ chọn những câu loại 1 lưu trước vào `df`.

In [7]:
col = df_1.columns
newcol = []
for i in col:
    if '_' not in i:
        newcol.append(i)

In [8]:
df = df_1[newcol]
df.head(5)

Duration (in seconds)     Q2     Q3           Q4   Q5  \
11                   470  55-59    Man  South Korea   No   
22                   174  25-29  Woman        China  Yes   
36                   506  25-29    Man        Japan   No   
71                   283  22-24  Woman        China  Yes   
72                   492  40-44    Man        Japan   No   

                                      Q8   Q9        Q11  \
11                       Master’s degree   No  1-3 years   
22                       Master’s degree   No  < 1 years   
36                       Master’s degree   No  3-5 years   
71                I prefer not to answer  NaN  < 1 years   
72  No formal education past high school  NaN  < 1 years   

                                      Q16  Q22             Q23  \
11                              1-2 years  NaN           Other   
22                                    NaN  NaN             NaN   
36                              2-3 years  NaN  Data Scientist   
71                           Under 1 year  NaN             NaN   
72  I do not use machine learning methods  NaN           Other   

                          Q24                       Q25  Q26  \
11  Manufacturing/Fabrication  10,000 or more employees  20+   
22                        NaN                       NaN  NaN   
36       Computers/Technology      1000-9,999 employees  1-2   
71                        NaN                       NaN  NaN   
72  Manufacturing/Fabrication      1000-9,999 employees    0   

                                                  Q27            Q29  \
11  We recently started using ML methods (i.e., mo...            NaN   
22                                                NaN            NaN   
36  We recently started using ML methods (i.e., mo...  50,000-59,999   
71                                                NaN            NaN   
72                                      I do not know  50,000-59,999   

                Q30  Q32    Q43  
11              NaN  NaN    NaN  
22              NaN  NaN    NaN  
36        $100-$999  NaN  Never  
71              NaN  NaN    NaN  
72  $10,000-$99,999  NaN    NaN

Tiếp đến, ta sẽ xử lý những câu hỏi loại 2 và tiếp tục lưu nó vào `df`

In [9]:
for i in range(len(col)):
    ij = col[i] + ' ' 
    if '_1 ' in ij:
        arr = [elem for elem in col if elem.startswith(col[i][0:col[i].index("_")])]
        temp_df = df_1[arr]
        a = [tuple(filter(lambda x: pd.notnull(x), row)) for row in temp_df.values]
        df[col[i][0:col[i].index("_")]] = a
df = df.applymap(lambda x: np.nan if isinstance(x, tuple) and len(x) == 0 else x)

C:\Users\KHUONG\AppData\Local\Temp\ipykernel_4100\3130294925.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col[i][0:col[i].index("_")]] = a


Vì name của columns đang ở dạng Q1, Q2, ... nên ta sẽ chuyển thành câu hỏi và reindex để dữ liệu dễ phân tích hơn.

In [10]:
sort_col = ['Duration (in seconds)']
for i in range(2,45):
    sort_col.append('Q' + str(i))
df = df.reindex(columns=sort_col)
df = df.reset_index(drop=True)
df.rename(columns=new_dic, inplace = True)
df.index = df.index + 1
df

Duration What is your age (# years)? What is your gender?  \
1         470                       55-59                  Man   
2         174                       25-29                Woman   
3         506                       25-29                  Man   
4         283                       22-24                Woman   
5         492                       40-44                  Man   
...       ...                         ...                  ...   
1534      156                       45-49                  Man   
1535    43922                       18-21                  Man   
1536      617                       50-54                  Man   
1537     1358                       18-21                  Man   
1538     1651                       35-39                  Man   

     In which country do you currently reside? Are you currently a student?  \
1                                  South Korea                           No   
2                                        China                          Yes   
3                                        Japan                           No   
4                                        China                          Yes   
5                                        Japan                           No   
...                                        ...                          ...   
1534                                     Japan                           No   
1535                                     Japan                          Yes   
1536                                     Japan                           No   
1537                                     China                          Yes   
1538                                     China                           No   

     On which platforms have you begun or completed data science courses?  \
1     (University Courses (resulting in a university...                     
2     (Kaggle Learn Courses, University Courses (res...                     
3                                              (Udemy,)                     
4     (University Courses (resulting in a university...                     
5                               (Kaggle Learn Courses,)                     
...                                                 ...                     
1534                            (Kaggle Learn Courses,)                     
1535  (University Courses (resulting in a university...                     
1536                                            (None,)                     
1537  (University Courses (resulting in a university...                     
1538                                        (Coursera,)                     

     What products or platforms did you find to be most helpful when you first started studying data science?  \
1              (Kaggle (notebooks, competitions, etc),)                                                         
2     (University courses, Online courses (Coursera,...                                                         
3                (Online courses (Coursera, EdX, etc),)                                                         
4     (University courses, Social media platforms (R...                                                         
5     (Video platforms (YouTube, Twitch, etc), Kaggl...                                                         
...                                                 ...                                                         
1534   (Social media platforms (Reddit, Twitter, etc),)                                                         
1535             (Online courses (Coursera, EdX, etc),)                                                         
1536           (Kaggle (notebooks, competitions, etc),)                                                         
1537  (Online courses (Coursera, EdX, etc), Social m...                                                         
1538             (Online courses (Coursera, EdX, etc),)                                        

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột

In [11]:
df.shape

(1538, 44)

Dữ liệu có 1538 dòng và 44 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Ta sẽ xem thử rằng trong tập có dòng nào bị trùng không. Nếu trùng thì trả về True, không thì là False và lưu số dòng trùng vào biến num_duplicated_rows

In [12]:
#Tính tổng dòng lặp
num_duplicated_rows = df.duplicated().sum()
is_duplicated = (num_duplicated_rows != 0)
print('So dong lap la {}'.format(num_duplicated_rows))
if is_duplicated == False:
    print('Bo du lieu khong lap')

So dong lap la 0
Bo du lieu khong lap


Vậy không có dòng lặp trong bộ dữ liệu này

### Mỗi cột có ý nghĩa gì?

In [13]:
df.columns

Index(['Duration', 'What is your age (# years)?', 'What is your gender?',
       'In which country do you currently reside?',
       'Are you currently a student?',
       'On which platforms have you begun or completed data science courses?',
       'What products or platforms did you find to be most helpful when you first started studying data science?',
       'What is the highest level of formal education that you have attained or plan to attain within the next 2 years?',
       'Have you ever published any academic research (papers, preprints, conference proceedings, etc)?',
       'Did your research make use of machine learning?',
       'For how many years have you been writing code and/or programming?',
       'What programming languages do you use on a regular basis?',
       'Which of the following integrated development environments (IDE's) do you use on a regular basis?',
       'Do you use any of the following hosted notebook products?',
       'Do you use any of the follo

Ý nghĩa của mỗi cột là:

- **Duration** là thời gian(s)  thực hiện khảo sát.

- Các cột còn lại là câu hỏi khảo sát. Và giá trị của các cột là câu trả lời.

Ta kiểm tra dạng dữ liệu của các cột để xem có đúng với dạng mà ta mong muốn không.

In [14]:
df.dtypes

Duration                                                                                                                                                      object
What is your age (# years)?                                                                                                                                   object
What is your gender?                                                                                                                                          object
In which country do you currently reside?                                                                                                                     object
Are you currently a student?                                                                                                                                  object
On which platforms have you begun or completed data science courses?                                                                                          object
What produ

Cột **Duration** đang ở dạng object và sai định dạng. Ta sẽ chuyển nó về dạng số `int64`

In [15]:
df['Duration'] = df['Duration'].astype('int64')

Ta thấy rằng các cột đã có định dạng đúng với định dạng mà ta mong muốn. Gồm 1 cột **Duration** có dạng **int** và toàn bộ cột còn lại có dạng **object**

## XỬ LÝ DỮ LIỆU NAN

Ta xem thử mỗi cột có bao nhiêu dữ liệu nan

In [16]:
df.isnull().sum()

Duration                                                                                                                                                         0
What is your age (# years)?                                                                                                                                      0
What is your gender?                                                                                                                                             0
In which country do you currently reside?                                                                                                                        0
Are you currently a student?                                                                                                                                     0
On which platforms have you begun or completed data science courses?                                                                                             0
What products or platf

- Với 6 câu hỏi (columns) đầu tiên, ta không thấy xuất hiện giá trị NAN. Có lẽ đây là những câu trả hỏi bắc buộc của khảo sát nên rất đầy đủ giá trị.

- Với các câu hỏi (columns) còn lại, ta thấy xuất hiện rất nhiều giá trị NAN. Vì đây là những câu hỏi tự chọn. Ta sẽ xoá những cột có quá nhiều giá trị NAN vì những cột này sẽ mang lại rất ít thông tin những lại gây khó khăn nhiều trong quá trình phân tích. Với dữ liệu hơn 1500 dòng, ta sẽ xoá những dòng có hơn 500 dòng NAN.

In [17]:
nan_counts = df.isna().sum()

columns_to_drop = nan_counts[nan_counts > 500].index

df.drop(columns=columns_to_drop, inplace= True)
df.isna().sum()

Duration                                                                                                             0
What is your age (# years)?                                                                                          0
What is your gender?                                                                                                 0
In which country do you currently reside?                                                                            0
Are you currently a student?                                                                                         0
On which platforms have you begun or completed data science courses?                                                 0
What products or platforms did you find to be most helpful when you first started studying data science?            27
What is the highest level of formal education that you have attained or plan to attain within the next 2 years?     66
For how many years have you been writing code an

Giờ dữ liệu chỉ còn những cột có ít hơn 500 dòng chứa NAN.

### Sự phân bố của dữ liệu

### Với mỗi cột có kiểu dữ liệu dạng số, các giá trị được phân bố như thế nào?

In [18]:
dtypes = df.dtypes
ind = dtypes[(dtypes == np.int64)].index
ind

Index(['Duration'], dtype='object')

In [19]:
df[ind]

Duration
1          470
2          174
3          506
4          283
5          492
...        ...
1534       156
1535     43922
1536       617
1537      1358
1538      1651

[1538 rows x 1 columns]

Ta có cột dữ liệu dạng số 'Duration'. Ta sẽ tính min, max, mean, median để xem sự phân bố của dữ liệu và có gì bất thường ở các cột này không?

In [20]:
print('Min:', int(df[ind].min()))
print('Max:', int(df[ind].max()))
print('Mean:', int(df[ind].mean()))
print('Median:', int(df[ind].median()))

Min: 120
Max: 2445778
Mean: 17530
Median: 507


Ta thấy các giá trị ở cột này không có gì bất thường

### Với mỗi cột có kiểu dữ liệu không phải dạng số, các giá trị được phân bố như thế nào?

In [21]:
dtypes = df.dtypes
ind = dtypes[(dtypes != np.int64)].index
ind

Index(['What is your age (# years)?', 'What is your gender?',
       'In which country do you currently reside?',
       'Are you currently a student?',
       'On which platforms have you begun or completed data science courses?',
       'What products or platforms did you find to be most helpful when you first started studying data science?',
       'What is the highest level of formal education that you have attained or plan to attain within the next 2 years?',
       'For how many years have you been writing code and/or programming?',
       'What programming languages do you use on a regular basis?',
       'Which of the following integrated development environments (IDE's) do you use on a regular basis?',
       'Do you use any of the following hosted notebook products?',
       'Do you use any of the following data visualization libraries on a regular basis?',
       'For how many years have you used machine learning methods?',
       'Which of the following machine learning fra

Ta có các cột dữ liệu dạng object như trên. Ta thấy rằng các cột dữ liệu là các câu hỏi lựa chọn. Ta sẽ tính số lượng giá trị, giá trị khác nhau, giá trị xuất hiện nhiều nhất và số lần xuất hiện để xem sự phân bố của dữ liệu và có gì bất thường ở các cột này không?

In [22]:
df[ind].describe()

What is your age (# years)? What is your gender?  \
count                         1538                 1538   
unique                          11                    4   
top                          25-29                  Man   
freq                           276                 1300   

       In which country do you currently reside? Are you currently a student?  \
count                                       1538                         1538   
unique                                         4                            2   
top                                        Japan                           No   
freq                                         556                          923   

       On which platforms have you begun or completed data science courses?  \
count                                                1538                     
unique                                                178                     
top                                               (None,)                     
freq                                                  318                     

       What products or platforms did you find to be most helpful when you first started studying data science?  \
count                                                1511                                                         
unique                                                 54                                                         
top              (Kaggle (notebooks, competitions, etc),)                                                         
freq                                                  219                                                         

       What is the highest level of formal education that you have attained or plan to attain within the next 2 years?  \
count                                                1472                                                                
unique                                                  7                                                                
top                                       Master’s degree                                                                
freq                                                  530                                                                

       For how many years have you been writing code and/or programming?  \
count                                                1456                  
unique                                                  7                  
top                                             1-3 years                  
freq                                                  432                  

       What programming languages do you use on a regular basis?  \
count                                                1387          
unique                                                329          
top                                             (Python,)          
freq                                                  323          

       Which of the following integrated development environments (IDE's) do you use on a regular basis?  \
count                                                1368                                                  
unique                                                365                                                  
top     ( Visual Studio Code (VSCode) ,  Jupyter Noteb...                                                  
freq                                                  100                                                  

       Do you use any of the following hosted notebook products?  \
count                                                1332          
unique                                                 97          
top                                               (None,)          
freq                                                  396          

       Do you use any of the following data visualization libraries on a regular basis?  \
count                      

Ta thấy không có gì bất thường ở các cột dữ liệu này

Qua quá trình khám phá và tiền xử lý dữ liệu. Ta thấy dữ liệu đã 'sạch' hơn. Ta sẽ lưu vào file `data_science_asia_clean.csv`.

In [23]:
df.to_csv('data_science_asia_clean.csv')